### PR TITLE
[merged] Libcurl prep

### PR DESCRIPTION
**Pull request by cgwalters**

### CODE REVIEW

**Comment by cgwalters** [comment]
I originally started porting things to use `char*` instead, then reversed myself and added `OstreeFetcherURI`.  In general the use of `GPtrArray` loses type safety for things like this, so review carefully.

The test coverage helped a lot, and the patch seems OK in self-review too.

**Comment by giuseppe** [comment]
LGTM, but I'll wait for @jlebon's review as he had recently touched this part

File: src/libostree/ostree-repo-pull.c [hunk 1]
```diff
@@ -2116,27 +2113,24 @@ fetch_mirrorlist (OstreeFetcher  *fetcher,
...
                                                               "config", NULL);
-          SoupURI *config_uri = soup_uri_new (config_uri_str);
+          g_autoptr(OstreeFetcherURI) config_uri = _ostree_fetcher_uri_parse (config_uri_str, NULL);
+
+          g_assert (config_uri);
 
           if (fetch_uri_contents_utf8_sync (fetcher, config_uri, NULL,
                                             cancellable, &local_error))
-            g_ptr_array_add (ret_mirrorlist, g_steal_pointer (&mirror_uri));
+            g_ptr_array_add (ret_mirrorlist, g_steal_pointer (&config_uri));
```
Comment (jlebon) [inline src/libostree/ostree-repo-pull.c]: But this should still be `mirror_uri`, right?

File: src/libostree/ostree-repo-pull.c [hunk 2]
```diff
@@ -2083,27 +2079,28 @@ fetch_mirrorlist (OstreeFetcher  *fetcher,
...
+      if (!mirror_uri)
         {
           g_debug ("Can't parse mirrorlist line '%s'", mirror_uri_str);
           continue;
         }
-      else if ((strcmp (soup_uri_get_scheme (mirror_uri), "http") != 0) &&
-               (strcmp (soup_uri_get_scheme (mirror_uri), "https") != 0))
+
+      scheme = _ostree_fetcher_uri_get_scheme (mirror_uri);
+      if ((strcmp (scheme, "http") != 0) && (strcmp (scheme, "https") != 0))
```
Comment (jlebon) [inline src/libostree/ostree-repo-pull.c]: `g_str_equal` would probably look nicer here.

File: src/libostree/ostree-repo-pull.c [hunk 3]
```diff
@@ -2111,15 +2111,11 @@ fetch_mirrorlist (OstreeFetcher  *fetcher,
       if (ret_mirrorlist->len == 0)
         {
           GError *local_error = NULL;
-          g_autofree char *config_uri_str = g_build_filename (mirror_uri_str,
-                                                              "config", NULL);
-          g_autoptr(OstreeFetcherURI) config_uri = _ostree_fetcher_uri_parse (config_uri_str, NULL);
-
-          g_assert (config_uri);
+          g_autoptr(OstreeFetcherURI) config_uri = _ostree_fetcher_uri_new_subpath (mirror_uri_str, "config");
```
Comment (jlebon) [inline src/libostree/ostree-repo-pull.c]: s/`mirror_uri_str`/`mirror_uri`/

**Comment by jlebon** [comment]
@rh-atomic-bot r+ 32e71fd

**Pull request closed without merging**